File: lib/imports/networking.py
from constructs import Construct
from cdktf_cdktf_provider_aws.vpc import Vpc
from cdktf_cdktf_provider_aws.subnet import Subnet
from cdktf_cdktf_provider_aws.internet_gateway import InternetGateway
from cdktf_cdktf_provider_aws.route_table import RouteTable
from cdktf_cdktf_provider_aws.route_table_association import RouteTableAssociation
from cdktf_cdktf_provider_aws.security_group import SecurityGroup, SecurityGroupIngress, SecurityGroupEgress
from cdktf_cdktf_provider_aws.vpc_peering_connection import VpcPeeringConnection
from cdktf_cdktf_provider_aws.vpc_peering_connection_accepter import VpcPeeringConnectionAccepterA
from cdktf_cdktf_provider_aws.route import Route


class NetworkingConstruct(Construct):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_suffix: str,
        primary_provider,
        secondary_provider
    ):
        super().__init__(scope, construct_id)

        self.environment_suffix = environment_suffix

        # Primary VPC (us-east-1)
        self.primary_vpc = Vpc(
            self,
            "primary_vpc",
            cidr_block="10.0.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"payment-primary-vpc-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        # Secondary VPC (us-west-2)
        self.secondary_vpc = Vpc(
            self,
            "secondary_vpc",
            cidr_block="10.1.0.0/16",
            enable_dns_hostnames=True,
            enable_dns_support=True,
            tags={
                "Name": f"payment-secondary-vpc-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

        # Primary region subnets (3 AZs)
        self.primary_private_subnets = []
        primary_azs = ["us-east-1a", "us-east-1b", "us-east-1c"]
        for i, az in enumerate(primary_azs):
            subnet = Subnet(
                self,
                f"primary_private_subnet_{i}",
                vpc_id=self.primary_vpc.id,
                cidr_block=f"10.0.{i}.0/24",
                availability_zone=az,
                tags={
                    "Name": f"payment-primary-private-{i}-{environment_suffix}",
                    "Environment": environment_suffix
                },
                provider=primary_provider
            )
            self.primary_private_subnets.append(subnet)

        # Secondary region subnets (3 AZs)
        self.secondary_private_subnets = []
        secondary_azs = ["us-west-2a", "us-west-2b", "us-west-2c"]
        for i, az in enumerate(secondary_azs):
            subnet = Subnet(
                self,
                f"secondary_private_subnet_{i}",
                vpc_id=self.secondary_vpc.id,
                cidr_block=f"10.1.{i}.0/24",
                availability_zone=az,
                tags={
                    "Name": f"payment-secondary-private-{i}-{environment_suffix}",
                    "Environment": environment_suffix
                },
                provider=secondary_provider
            )
            self.secondary_private_subnets.append(subnet)

        # Internet Gateway for primary VPC
        self.primary_igw = InternetGateway(
            self,
            "primary_igw",
            vpc_id=self.primary_vpc.id,
            tags={
                "Name": f"payment-primary-igw-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        # Internet Gateway for secondary VPC
        self.secondary_igw = InternetGateway(
            self,
            "secondary_igw",
            vpc_id=self.secondary_vpc.id,
            tags={
                "Name": f"payment-secondary-igw-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

        # Route tables for primary VPC
        self.primary_route_table = RouteTable(
            self,
            "primary_route_table",
            vpc_id=self.primary_vpc.id,
            tags={
                "Name": f"payment-primary-rt-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        # Route tables for secondary VPC
        self.secondary_route_table = RouteTable(
            self,
            "secondary_route_table",
            vpc_id=self.secondary_vpc.id,
            tags={
                "Name": f"payment-secondary-rt-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

        # VPC Peering Connection
        self.vpc_peering = VpcPeeringConnection(
            self,
            "vpc_peering",
            vpc_id=self.primary_vpc.id,
            peer_vpc_id=self.secondary_vpc.id,
            peer_region="us-west-2",
            auto_accept=False,
            tags={
                "Name": f"payment-vpc-peering-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        # Accept VPC Peering in secondary region
        self.vpc_peering_accepter = VpcPeeringConnectionAccepterA(
            self,
            "vpc_peering_accepter",
            vpc_peering_connection_id=self.vpc_peering.id,
            auto_accept=True,
            tags={
                "Name": f"payment-vpc-peering-accepter-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

        # Routes for VPC peering in primary VPC
        Route(
            self,
            "primary_peering_route",
            route_table_id=self.primary_route_table.id,
            destination_cidr_block="10.1.0.0/16",
            vpc_peering_connection_id=self.vpc_peering.id,
            provider=primary_provider
        )

        # Routes for VPC peering in secondary VPC
        Route(
            self,
            "secondary_peering_route",
            route_table_id=self.secondary_route_table.id,
            destination_cidr_block="10.0.0.0/16",
            vpc_peering_connection_id=self.vpc_peering.id,
            provider=secondary_provider
        )

        # Associate subnets with route tables
        for i, subnet in enumerate(self.primary_private_subnets):
            RouteTableAssociation(
                self,
                f"primary_rt_assoc_{i}",
                subnet_id=subnet.id,
                route_table_id=self.primary_route_table.id,
                provider=primary_provider
            )

        for i, subnet in enumerate(self.secondary_private_subnets):
            RouteTableAssociation(
                self,
                f"secondary_rt_assoc_{i}",
                subnet_id=subnet.id,
                route_table_id=self.secondary_route_table.id,
                provider=secondary_provider
            )

        # Security Groups - Primary Region
        self.primary_db_sg = SecurityGroup(
            self,
            "primary_db_sg",
            name=f"payment-primary-db-sg-{environment_suffix}",
            description="Security group for Aurora database in primary region",
            vpc_id=self.primary_vpc.id,
            ingress=[
                SecurityGroupIngress(
                    from_port=5432,
                    to_port=5432,
                    protocol="tcp",
                    cidr_blocks=["10.0.0.0/16", "10.1.0.0/16"]
                )
            ],
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"]
                )
            ],
            tags={
                "Name": f"payment-primary-db-sg-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        self.primary_lambda_sg = SecurityGroup(
            self,
            "primary_lambda_sg",
            name=f"payment-primary-lambda-sg-{environment_suffix}",
            description="Security group for Lambda functions in primary region",
            vpc_id=self.primary_vpc.id,
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"]
                )
            ],
            tags={
                "Name": f"payment-primary-lambda-sg-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=primary_provider
        )

        # Security Groups - Secondary Region
        self.secondary_db_sg = SecurityGroup(
            self,
            "secondary_db_sg",
            name=f"payment-secondary-db-sg-{environment_suffix}",
            description="Security group for Aurora database in secondary region",
            vpc_id=self.secondary_vpc.id,
            ingress=[
                SecurityGroupIngress(
                    from_port=5432,
                    to_port=5432,
                    protocol="tcp",
                    cidr_blocks=["10.0.0.0/16", "10.1.0.0/16"]
                )
            ],
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"]
                )
            ],
            tags={
                "Name": f"payment-secondary-db-sg-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

        self.secondary_lambda_sg = SecurityGroup(
            self,
            "secondary_lambda_sg",
            name=f"payment-secondary-lambda-sg-{environment_suffix}",
            description="Security group for Lambda functions in secondary region",
            vpc_id=self.secondary_vpc.id,
            egress=[
                SecurityGroupEgress(
                    from_port=0,
                    to_port=0,
                    protocol="-1",
                    cidr_blocks=["0.0.0.0/0"]
                )
            ],
            tags={
                "Name": f"payment-secondary-lambda-sg-{environment_suffix}",
                "Environment": environment_suffix
            },
            provider=secondary_provider
        )

    @property
    def primary_vpc_id(self):
        return self.primary_vpc.id

    @property
    def secondary_vpc_id(self):
        return self.secondary_vpc.id

    @property
    def primary_private_subnet_ids(self):
        return [subnet.id for subnet in self.primary_private_subnets]

    @property
    def secondary_private_subnet_ids(self):
        return [subnet.id for subnet in self.secondary_private_subnets]

    @property
    def primary_db_sg_id(self):
        return self.primary_db_sg.id

    @property
    def secondary_db_sg_id(self):
        return self.secondary_db_sg.id

    @property
    def primary_lambda_sg_id(self):
        return self.primary_lambda_sg.id

    @property
    def secondary_lambda_sg_id(self):
        return self.secondary_lambda_sg.id

File: archive/cdktf-ts/Pr7779/lib/constructs/networking.ts
import { Construct } from 'constructs';
import { AwsProvider } from '@cdktf/provider-aws/lib/provider';
import { Vpc } from '@cdktf/provider-aws/lib/vpc';
import { Subnet } from '@cdktf/provider-aws/lib/subnet';
import { InternetGateway } from '@cdktf/provider-aws/lib/internet-gateway';
import { RouteTable } from '@cdktf/provider-aws/lib/route-table';
import { RouteTableAssociation } from '@cdktf/provider-aws/lib/route-table-association';
import { Route } from '@cdktf/provider-aws/lib/route';
import { SecurityGroup } from '@cdktf/provider-aws/lib/security-group';
import { SecurityGroupRule } from '@cdktf/provider-aws/lib/security-group-rule';

export interface NetworkingConstructProps {
  environmentSuffix: string;
  primaryProvider: AwsProvider;
  secondaryProvider: AwsProvider;
}

export class NetworkingConstruct extends Construct {
  public readonly primaryVpcId: string;
  public readonly secondaryVpcId: string;
  public readonly primaryPrivateSubnetIds: string[];
  public readonly secondaryPrivateSubnetIds: string[];
  public readonly primaryDbSecurityGroupId: string;
  public readonly secondaryDbSecurityGroupId: string;
  public readonly primaryLambdaSecurityGroupId: string;
  public readonly secondaryLambdaSecurityGroupId: string;

  constructor(scope: Construct, id: string, props: NetworkingConstructProps) {
    super(scope, id);

    const { environmentSuffix, primaryProvider, secondaryProvider } = props;

    // Primary Region VPC
    const primaryVpc = new Vpc(this, 'PrimaryVPC', {
      provider: primaryProvider,
      cidrBlock: '10.0.0.0/16',
      enableDnsHostnames: true,
      enableDnsSupport: true,
      tags: {
        Name: `dr-vpc-primary-${environmentSuffix}`,
      },
    });

    // Secondary Region VPC
    const secondaryVpc = new Vpc(this, 'SecondaryVPC', {
      provider: secondaryProvider,
      cidrBlock: '10.1.0.0/16',
      enableDnsHostnames: true,
      enableDnsSupport: true,
      tags: {
        Name: `dr-vpc-secondary-${environmentSuffix}`,
      },
    });

    // Primary Region - Internet Gateway
    const primaryIgw = new InternetGateway(this, 'PrimaryIGW', {
      provider: primaryProvider,
      vpcId: primaryVpc.id,
      tags: {
        Name: `dr-igw-primary-${environmentSuffix}`,
      },
    });

    // Secondary Region - Internet Gateway
    const secondaryIgw = new InternetGateway(this, 'SecondaryIGW', {
      provider: secondaryProvider,
      vpcId: secondaryVpc.id,
      tags: {
        Name: `dr-igw-secondary-${environmentSuffix}`,
      },
    });

    // Primary Region - Private Subnets (for Lambda and RDS)
    const primaryPrivateSubnet1 = new Subnet(this, 'PrimaryPrivateSubnet1', {
      provider: primaryProvider,
      vpcId: primaryVpc.id,
      cidrBlock: '10.0.1.0/24',
      availabilityZone: 'us-east-1a',
      tags: {
        Name: `dr-private-subnet-1-primary-${environmentSuffix}`,
      },
    });

    const primaryPrivateSubnet2 = new Subnet(this, 'PrimaryPrivateSubnet2', {
      provider: primaryProvider,
      vpcId: primaryVpc.id,
      cidrBlock: '10.0.2.0/24',
      availabilityZone: 'us-east-1b',
      tags: {
        Name: `dr-private-subnet-2-primary-${environmentSuffix}`,
      },
    });

    // Primary Region - Public Subnets (for NAT gateway alternatives or public access)
    const primaryPublicSubnet1 = new Subnet(this, 'PrimaryPublicSubnet1', {
      provider: primaryProvider,
      vpcId: primaryVpc.id,
      cidrBlock: '10.0.10.0/24',
      availabilityZone: 'us-east-1a',
      mapPublicIpOnLaunch: true,
      tags: {
        Name: `dr-public-subnet-1-primary-${environmentSuffix}`,
      },
    });

    // Primary Region - Route Table for Public Subnets
    const primaryPublicRouteTable = new RouteTable(this, 'PrimaryPublicRT', {
      provider: primaryProvider,
      vpcId: primaryVpc.id,
      tags: {
        Name: `dr-public-rt-primary-${environmentSuffix}`,
      },
    });

    new Route(this, 'PrimaryPublicRoute', {
      provider: primaryProvider,
      routeTableId: primaryPublicRouteTable.id,
      destinationCidrBlock: '0.0.0.0/0',
      gatewayId: primaryIgw.id,
    });

    new RouteTableAssociation(this, 'PrimaryPublicRTAssoc', {
      provider: primaryProvider,
      subnetId: primaryPublicSubnet1.id,
      routeTableId: primaryPublicRouteTable.id,
    });

    // Secondary Region - Private Subnets
    const secondaryPrivateSubnet1 = new Subnet(
      this,
      'SecondaryPrivateSubnet1',
      {
        provider: secondaryProvider,
        vpcId: secondaryVpc.id,
        cidrBlock: '10.1.1.0/24',
        availabilityZone: 'us-east-2a',
        tags: {
          Name: `dr-private-subnet-1-secondary-${environmentSuffix}`,
        },
      }
    );

    const secondaryPrivateSubnet2 = new Subnet(
      this,
      'SecondaryPrivateSubnet2',
      {
        provider: secondaryProvider,
        vpcId: secondaryVpc.id,
        cidrBlock: '10.1.2.0/24',
        availabilityZone: 'us-east-2b',
        tags: {
          Name: `dr-private-subnet-2-secondary-${environmentSuffix}`,
        },
      }
    );

    // Secondary Region - Public Subnets
    const secondaryPublicSubnet1 = new Subnet(this, 'SecondaryPublicSubnet1', {
      provider: secondaryProvider,
      vpcId: secondaryVpc.id,
      cidrBlock: '10.1.10.0/24',
      availabilityZone: 'us-east-2a',
      mapPublicIpOnLaunch: true,
      tags: {
        Name: `dr-public-subnet-1-secondary-${environmentSuffix}`,
      },
    });

    // Secondary Region - Route Table for Public Subnets
    const secondaryPublicRouteTable = new RouteTable(
      this,
      'SecondaryPublicRT',
      {
        provider: secondaryProvider,
        vpcId: secondaryVpc.id,
        tags: {
          Name: `dr-public-rt-secondary-${environmentSuffix}`,
        },
      }
    );

    new Route(this, 'SecondaryPublicRoute', {
      provider: secondaryProvider,
      routeTableId: secondaryPublicRouteTable.id,
      destinationCidrBlock: '0.0.0.0/0',
      gatewayId: secondaryIgw.id,
    });

    new RouteTableAssociation(this, 'SecondaryPublicRTAssoc', {
      provider: secondaryProvider,
      subnetId: secondaryPublicSubnet1.id,
      routeTableId: secondaryPublicRouteTable.id,
    });

    // Primary Region - Security Group for RDS
    const primaryDbSecurityGroup = new SecurityGroup(
      this,
      'PrimaryDBSecurityGroup',
      {
        provider: primaryProvider,
        vpcId: primaryVpc.id,
        name: `dr-db-sg-primary-${environmentSuffix}`,
        description: 'Security group for Aurora database in primary region',
        tags: {
          Name: `dr-db-sg-primary-${environmentSuffix}`,
        },
      }
    );

    // Allow Lambda to access RDS on port 5432
    new SecurityGroupRule(this, 'PrimaryDBIngressFromLambda', {
      provider: primaryProvider,
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      securityGroupId: primaryDbSecurityGroup.id,
      sourceSecurityGroupId: primaryDbSecurityGroup.id,
      description: 'Allow PostgreSQL access from Lambda',
    });

    new SecurityGroupRule(this, 'PrimaryDBEgress', {
      provider: primaryProvider,
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: primaryDbSecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    // Secondary Region - Security Group for RDS
    const secondaryDbSecurityGroup = new SecurityGroup(
      this,
      'SecondaryDBSecurityGroup',
      {
        provider: secondaryProvider,
        vpcId: secondaryVpc.id,
        name: `dr-db-sg-secondary-${environmentSuffix}`,
        description: 'Security group for Aurora database in secondary region',
        tags: {
          Name: `dr-db-sg-secondary-${environmentSuffix}`,
        },
      }
    );

    new SecurityGroupRule(this, 'SecondaryDBIngressFromLambda', {
      provider: secondaryProvider,
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      securityGroupId: secondaryDbSecurityGroup.id,
      sourceSecurityGroupId: secondaryDbSecurityGroup.id,
      description: 'Allow PostgreSQL access from Lambda',
    });

    new SecurityGroupRule(this, 'SecondaryDBEgress', {
      provider: secondaryProvider,
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: secondaryDbSecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    // Primary Region - Security Group for Lambda
    const primaryLambdaSecurityGroup = new SecurityGroup(
      this,
      'PrimaryLambdaSecurityGroup',
      {
        provider: primaryProvider,
        vpcId: primaryVpc.id,
        name: `dr-lambda-sg-primary-${environmentSuffix}`,
        description: 'Security group for Lambda functions in primary region',
        tags: {
          Name: `dr-lambda-sg-primary-${environmentSuffix}`,
        },
      }
    );

    new SecurityGroupRule(this, 'PrimaryLambdaEgress', {
      provider: primaryProvider,
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: primaryLambdaSecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    // Allow Lambda to access RDS
    new SecurityGroupRule(this, 'PrimaryLambdaToDBIngress', {
      provider: primaryProvider,
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      securityGroupId: primaryDbSecurityGroup.id,
      sourceSecurityGroupId: primaryLambdaSecurityGroup.id,
      description: 'Allow Lambda to access Aurora',
    });

    // Secondary Region - Security Group for Lambda
    const secondaryLambdaSecurityGroup = new SecurityGroup(
      this,
      'SecondaryLambdaSecurityGroup',
      {
        provider: secondaryProvider,
        vpcId: secondaryVpc.id,
        name: `dr-lambda-sg-secondary-${environmentSuffix}`,
        description: 'Security group for Lambda functions in secondary region',
        tags: {
          Name: `dr-lambda-sg-secondary-${environmentSuffix}`,
        },
      }
    );

    new SecurityGroupRule(this, 'SecondaryLambdaEgress', {
      provider: secondaryProvider,
      type: 'egress',
      fromPort: 0,
      toPort: 0,
      protocol: '-1',
      cidrBlocks: ['0.0.0.0/0'],
      securityGroupId: secondaryLambdaSecurityGroup.id,
      description: 'Allow all outbound traffic',
    });

    new SecurityGroupRule(this, 'SecondaryLambdaToDBIngress', {
      provider: secondaryProvider,
      type: 'ingress',
      fromPort: 5432,
      toPort: 5432,
      protocol: 'tcp',
      securityGroupId: secondaryDbSecurityGroup.id,
      sourceSecurityGroupId: secondaryLambdaSecurityGroup.id,
      description: 'Allow Lambda to access Aurora',
    });

    // Export values
    this.primaryVpcId = primaryVpc.id;
    this.secondaryVpcId = secondaryVpc.id;
    this.primaryPrivateSubnetIds = [
      primaryPrivateSubnet1.id,
      primaryPrivateSubnet2.id,
    ];
    this.secondaryPrivateSubnetIds = [
      secondaryPrivateSubnet1.id,
      secondaryPrivateSubnet2.id,
    ];
    this.primaryDbSecurityGroupId = primaryDbSecurityGroup.id;
    this.secondaryDbSecurityGroupId = secondaryDbSecurityGroup.id;
    this.primaryLambdaSecurityGroupId = primaryLambdaSecurityGroup.id;
    this.secondaryLambdaSecurityGroupId = secondaryLambdaSecurityGroup.id;
  }
}
